api/eds.proto

```diff
@@ -0,0 +1,183 @@
+  // TODO(htuch): Is there some restriction that all weights in a cluster must
+  // sum to 100? Do we want to be more specific and have per-zone and
+  // per-subzone as distinct weights for example? Can the weight also be 0?
+  google.protobuf.UInt32Value load_balancing_weight = 3;
+}
+
+message ClusterEndpoints {
+  string cluster_name = 1;
+  repeated LocalityEndpoints locality_endpoints = 2;
+}
+
+message EndpointDiscoveryResponse {
+  repeated ClusterEndpoints cluster_endpoints = 1;
+}
+
+// Example load report from a single request:
+//
+// [metric name, metric value]
+// * cpu_seconds, 0.7
+// * flash_utilization, 75
+//
+// When aggregating Envoy needs to count how many request's load reports
+// included each metric type, so Envoy can account for requests that don't
+// include that metric type. e.g.:
+//
+// [name, count, sum of values]
+// * cpu_seconds, 10, 17.5
+// * flash_utilization, 5, 375
+message EndpointLoadMetricStats {
+  // Name of the metric; may be empty.
+  string metric_name = 1;
+  // Number of calls that finished and included this metric.
+  uint64 num_requests_finished_with_metric = 2;
+  // Sum of metric values across all calls that finished with this metric for
+  // load_reporting_interval.
+  double total_metric_value = 3;
+}
+
+// These are stats Envoy reports to GLB every so often. Report frequency is
+// defined by LoadBalanceResponse.interval
+// Stats per upstream region/zone and optionally per subzone
+message UpstreamLocalityStats {
+  // Name of zone, region and optionally endpoint group this metrics was
+  // collected from. Zone and region names could be empty if unknown.
+  string Locality = 1;
+
+  // The total number of requests sent by this Envoy since the last report.
+  uint64 total_requests = 2;
+  // The total number of unfinished requests
+  uint64 total_requests_in_progress = 3;
+  // The number of errors since the last report.
+  // TODO(htuch): Ideally we would have the tcp_errors key be TcpErrorType, but
+  // enums are not supported as map key types.
+  enum TcpErrorType {
+    TIMEOUT = 0;
+    // TODO(htuch): Fill in additional TcpErrorType values.
+  }
+  map<uint32, uint64> tcp_errors = 4;
+  // HTTP status code, count
+  map<uint32, uint64> http_errors = 5;
+  // GRCP status code, count
+  map<uint32, uint64> grpc_errors = 6;
+  // The number of dropped requests since the last report.
+  uint64 dropped_requests = 7;
+
+  // Stats for multi-dimensional load balancing.
+  repeated EndpointLoadMetricStats load_metric_stats = 8;
+}
+
+// Per cluster stats
+message ClusterStats {
+  string cluster_name = 1;
+  // Need at least one.
+  repeated UpstreamLocalityStats upstream_locality_stats = 2;
+}
+
+message LoadBalanceRequest {
+  Node node = 1;  // zone/region where this Envoy runs
+  repeated ClusterStats cluster_stats = 3;
+}
+
+message ClusterLoadBalance {
+  string cluster_name = 1;
+  repeated LocalityEndpoints endpoints = 2;
+  // Add equivalent of drop overload.
```

Comment:
drop_overload (or may be we should call it drop_ratio) is the fraction of traffic that needs to be dropped. This is admin-like action that allows to protect upstream hosts should they unable to recover from outage or should they are unable to autoscale and hence overall incoming traffic volume need to be trimmed to protect them.

failover_endpoints. In case all endpoints for a particular zone/subzone is unavailable/unhealthy, additional endpoints are given out for use in case of catastrophic failure. They also have weights.

drop overload and failover are not related.